src/OrbitBase/include/OrbitBase/Typedef.h

```diff
@@ -0,0 +1,127 @@
+// const Wrapper<int> kFirstWrapped(1);
+// const Wrapper<int> kSecondWrapped(2);
+// ```
+// and define a callable (not necessarily a lambda)
+// `auto add = [](int i, int j) { return i + j; };`
+// Finally, we can use `Apply` function to apply `add` to the values stored in `kFirstWrapped` and
+// `kSecondWrapped`. The returned value of the add will be wrapped in a typedef with the same
+// `Tag_`.
+// `const Wrapper<int> sum_wrapped = Apply(add, kFirstWrapped, kSecondWrapped);`
+// Naturally, had we supplied the arguments wrapped in Typedefs of different `Tag_`, the code
+// wouldn't compile.
+// ```
+// struct OtherTag {};
+// using OtherWrapper = Typedef<OtherTag, T>;
+// const OtherWrapper<int> kWrong(2);
+// Apply(add, kFirstWrapped, kWrong); // ERROR!!!
+// ```
+// See TypedefTest.cpp for more examples.
+template <typename Tag_, typename T>
+class Typedef {
+ public:
+  using Tag = Tag_;
+
+  [[nodiscard]] constexpr const T* operator->() const { return &value_; }
+  [[nodiscard]] constexpr T* operator->() { return &value_; }
+
+  [[nodiscard]] constexpr const T& operator*() const& { return value_; }
+  [[nodiscard]] constexpr T&& operator*() && { return std::move(value_); }
+  [[nodiscard]] constexpr const T&& operator*() const&& { return std::move(value_); }
+  [[nodiscard]] constexpr T& operator*() & { return value_; }
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  constexpr explicit Typedef(U&& value) : value_(std::forward<U>(value)) {}
+
+  template <typename... Args>
+  constexpr explicit Typedef(std::in_place_t, Args&&... args)
+      : value_(T(std::forward<T>(args)...)) {}
+
+  constexpr Typedef(Typedef&& other) = default;
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  constexpr Typedef(const Typedef<Tag_, U>& other) : value_(*other) {}
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  constexpr Typedef(Typedef<Tag_, U>&& other) : value_(std::move(*other)) {}
+
+  template <typename U, typename = EnableIfUIsNotConvertibleToT<T, U>, typename = void>
+  constexpr explicit Typedef(const Typedef<Tag_, U>& other) : value_(*other) {}
+
+  template <typename U, typename = EnableIfUIsNotConvertibleToT<T, U>, typename = void>
+  constexpr explicit Typedef(Typedef<Tag_, U>&& other) : value_(std::move(*other)) {}
+
+  constexpr Typedef& operator=(Typedef const&) = default;
+  constexpr Typedef& operator=(Typedef&&) = default;
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  constexpr Typedef& operator=(const Typedef<Tag_, U>& other) {
+    value_ = *other;
+    return *this;
+  }
+
+  template <typename U, typename = EnableIfUConvertibleToT<T, U>>
+  constexpr Typedef& operator=(Typedef<Tag_, U>&& other) {
+    value_ = std::move(*other);
+    return *this;
+  }
+
+ private:
+  T value_;
+};
+
+template <typename Tag>
+class Typedef<Tag, void> {};
+
+// `action` is a callable. `args` are its args wrapped in `Typedef`s  with the same `Tag`. The args
```

Comment:
> its args wrapped in `Typedef`s with the same `Tag`

This seems arbitrary. If two `Typedef`s wrap the same type, why would it not be allowed to call a function on them? Either `Apply` is too generic as a name (I think this is the case here), or the choice is arbitrary for a specific use-case.

Reply:
I don't think it's arbitrary. You don't want smth like `Time total = Apply(sum, duration1, scope_id)` to compile. 
What actually happens here: we _lift_ the function into container Timestamp<Tag, T> and apply it. Perhaps. `LiftAndApply` would be a better name, but _lift_ is FP terminology.

Reply:
Moreover, I DO want to allow for calling a function on Typedefs of the same Tag, yet of different T. This is already used in Mizar.

Reply:
In relation to the name, it's arbitrary. You could have a function that takes a duration and a scope id. We need a better name.

Reply:
Are you OK with LiftAndApply?

Reply:
Yeah, I can't think of anything better.

Reply:
@beckerhe are you OK with `LiftAndApply` ?

Reply:
yes